### PR TITLE
Make `isExpensiveLoadOrStore` consider blocked pointers load and stores

### PR DESCRIPTION
**Pull request by etiotto**
The `isExpensiveLoadOrStore` function (third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp) fails to consider block pointers and consequently always returns `false`  for loads (and stores) operations that use a block pointer. 
In turn, this causes  the `RemoveLayoutConversion` pass to never consider loads using block pointers as `anchor` operations. 

This PR changes `isExpensiveLoadOrStore` so that block pointer loads can be properly recognized. The `RemoveLayourConversion` pass is then able to consider those loads as anchor operations and preserve their layout. 

Because `RemoveLayoutConversion` is invoked at several points in the optimization pipeline, the change in  third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp alone causes performance degradation in a couple of GEMM like benchmarks, specifically when operand A  of `tl.dot` is transposed and when the input of `tl.dot` is first fed into an exponential.

These 2 performance degradation have ben fixed by an enhancing the `MaterializeBlockPointer` and `MatmulLoopPipeline` optimizations, so that they can retrieve the dot layout of block pointer loads transitively from its users (in those benchmarks the blocked layout of block ptrs loads is transitively converted to a dot layout).

### CODE REVIEW

File: test/TritonIntelGPU/backward_combine_dpas_dot_layout.mlir [hunk 1]
```diff
@@ -47,16 +47,16 @@ module attributes {"triton_gpu.num-warps" = 4 : i32, "triton_gpu.threads-per-war
...
     %22 = tt.make_tensor_ptr %arg1, [%16, %20], [%21, %c1_i64], [%c0_i32, %19] {order = array<i32: 1, 0>} : <tensor<32x256xf16, #blocked1>>
     // CHECK: %[[VAL_41:.*]]:3 = scf.for %{{.*}} = %{{.*}} to %{{.*}} step %{{.*}} iter_args(%{{.*}} = %{{.*}}, %{{.*}} = %[[VAL_36]], %{{.*}} = %[[VAL_40]]) -> (tensor<64x256xf32, #[[DPAS]]>, !tt.ptr<tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #[[DPAS]], kWidth = 2}>>>, !tt.ptr<tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #[[DPAS]], kWidth = 2}>>>)  : i32 {
-    // CHECK: %[[VAL_46:.*]] = tt.load %{{.*}} {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #[[DPAS]], kWidth = 2}>>>
-    // CHECK: %[[VAL_47:.*]] = tt.load %{{.*}} {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #[[DPAS]], kWidth = 2}>>>
+    // CHECK: %[[VAL_46:.*]] = tt.load %{{.*}} {boundaryCheck = array<i32: 0, 1>, triton_intel_gpu.block_io = "row_major"} : !tt.ptr<tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #[[DPAS]], kWidth = 2}>>>
```
Comment (etiotto) [inline test/TritonIntelGPU/backward_combine_dpas_dot_layout.mlir:50]: Note: adding `triton_intel_gpu.block_io` is consistent with our optimization pipeline (in our pipeline this is done before the 2nd invocation of `RemoveLayoutConversion`)

File: test/TritonIntelGPU/combine.mlir [hunk 2]
```diff
@@ -2324,31 +2324,29 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32
     %cst_1 = arith.constant dense<0.000000e+00> : tensor<256x256xf32, #blocked2>
     %0 = tt.get_program_id x : i32
     %1 = tt.get_program_id y : i32
-    // CHECK:   %[[VAL_0:.*]] = tt.make_tensor_ptr {{.*}} : <tensor<256x32xbf16, #triton_gpu.dot_op<{opIdx = 0, parent = #[[$DPAS]], kWidth = 2}>>>
-    // CHECK:   %[[VAL_1:.*]] = tt.make_tensor_ptr {{.*}} : <tensor<32x256xbf16, #triton_gpu.dot_op<{opIdx = 1, parent = #[[$DPAS]], kWidth = 2}>>>
+    // CHECK:   %[[VAL_0:.*]] = tt.make_tensor_ptr {{.*}} : <tensor<256x32xbf16, {{.*}}>>
```
Comment (etiotto) [inline test/TritonIntelGPU/combine.mlir:2327]: The actual layout is not important in these tests.

**Comment by etiotto** [comment]
@chengjunlu can you do a code review please ?

**Review by chengjunlu — APPROVED** [review]

**Comment by chengjunlu** [comment]
> @chengjunlu can you do a code review please ?

LGTM.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**